### PR TITLE
ETCD-583: add signer expiration metric and alert

### PR DESCRIPTION
**Pull request by tjungblu**
/hold

need to add an alert as well

### CODE REVIEW

**Comment by tjungblu** [comment]
/hold cancel

good to review

File: manifests/0000_90_etcd-operator_03_prometheusrule.yaml [hunk 1]
```diff
@@ -112,7 +112,7 @@ spec:
         runbook_url: https://etcd.io/docs/v3.5/op-guide/maintenance/#defragmentation
         summary: etcd database size in use is less than 50% of the actual allocated storage.
       expr: |
-        (last_over_time(etcd_mvcc_db_total_size_in_use_in_bytes[5m]) / last_over_time(etcd_mvcc_db_total_size_in_bytes[5m])) < 0.5 and etcd_mvcc_db_total_size_in_use_in_bytes > 104857600
+        (last_over_time(etcd_mvcc_db_total_size_in_use_in_bytes{job=~".*etcd.*"}[5m]) / last_over_time(etcd_mvcc_db_total_size_in_bytes{job=~".*etcd.*"}[5m])) < 0.5 and etcd_mvcc_db_total_size_in_use_in_bytes{job=~".*etcd.*"} > 104857600
```
Comment (tjungblu) [inline manifests/0000_90_etcd-operator_03_prometheusrule.yaml:115]: this is not harmful btw, that came through with some upstream changes. The metrics all contain the job label, just verified this

**Comment by tjungblu** [comment]
alerts will look like this:

image

not sure if we can round the integer somehow, otherwise that's doing the job  - provided we agree on the thresholds

File: pkg/operator/etcdcertsigner/etcdcertsignercontroller.go [hunk 2]
```diff
@@ -220,6 +237,8 @@ func (c *EtcdCertSignerController) syncAllMasterCertificates(ctx context.Context
 		return err
 	}
 
+	c.reportExpirationMetric(signerCaPair, "metrics-signer-ca")
```
Comment (hasbro17) [inline pkg/operator/etcdcertsigner/etcdcertsignercontroller.go:240]:
```suggestion
	c.reportExpirationMetric(metricsSignerCaPair, "metrics-signer-ca")
```

File: pkg/operator/etcdcertsigner/etcdcertsignercontroller.go [hunk 3]
```diff
@@ -126,17 +134,24 @@ func NewEtcdCertSignerController(
 		metricsClientCert:     metricsClientCert,
 	}
 
+	signerExpirationGauge := metrics.NewGaugeVec(&metrics.GaugeOpts{
+		Name: signerExpirationMetricName,
+		Help: "Report observed days to expiration (NOT_AFTER) of a given signer certificate over time",
```
Comment (hasbro17) [inline pkg/operator/etcdcertsigner/etcdcertsignercontroller.go:139]: Just to clarify what NOT_AFTER means here, i.e not after expiration? Or is this an env var of sorts?

**Review by hasbro17 — COMMENTED** [review]
Nice work, mostly LGTM. Couple of nits.

File: pkg/operator/etcdcertsigner/etcdcertsignercontroller.go (hunk 2 above)
Comment (tjungblu) [reply pkg/operator/etcdcertsigner/etcdcertsignercontroller.go:240]: ohh good catch

File: pkg/operator/etcdcertsigner/etcdcertsignercontroller.go (hunk 3 above)
Comment (tjungblu) [reply pkg/operator/etcdcertsigner/etcdcertsignercontroller.go:139]: it's certificate lingo: https://serverfault.com/a/339471 - but I think we can just leave this out of the help text

File: pkg/operator/etcdcertsigner/etcdcertsignercontroller.go (hunk 2 above)
Comment (tjungblu) [reply pkg/operator/etcdcertsigner/etcdcertsignercontroller.go]: fixed!

**Comment by hasbro17** [comment]
/override ci/prow/e2e-aws-ovn-etcd-scaling

Scaling test is passing alright, the failure seems unrelated.
`event should not see excessive RequiredInstallerResourcesMissing secrets `

**Pull request merged**